library_generation/generate_pr_description.py

```diff
@@ -0,0 +1,197 @@
+    baseline_commit. Only commits which change files in a pre-defined
+    file paths will be considered.
+    Note that baseline_commit should be an ancestor of latest_commit.
+
+    :param repo_url: the url of the repository.
+    :param latest_commit: the newest commit to be considered in
+    selecting commit message.
+    :param baseline_commit: the oldest commit to be considered in
+    selecting commit message. This commit should be an ancestor of
+    :param paths: a mapping from file paths to library_name.
+    :param generator_version: the version of the generator.
+    :param is_monorepo: whether to generate commit messages in a monorepo.
+    :return: commit messages.
+    """
+    tmp_dir = "/tmp/repo"
+    shutil.rmtree(tmp_dir, ignore_errors=True)
+    os.mkdir(tmp_dir)
+    repo = Repo.clone_from(repo_url, tmp_dir)
+    commit = repo.commit(latest_commit)
+    qualified_commits = {}
+    while str(commit.hexsha) != baseline_commit:
+        commit_and_name = __filter_qualified_commit(paths=paths, commit=commit)
+        if commit_and_name != ():
+            qualified_commits[commit_and_name[0]] = commit_and_name[1]
+        commit_parents = commit.parents
+        if len(commit_parents) == 0:
+            break
+        commit = commit_parents[0]
+    shutil.rmtree(tmp_dir, ignore_errors=True)
+    return __combine_commit_messages(
+        latest_commit=latest_commit,
+        baseline_commit=baseline_commit,
+        commits=qualified_commits,
+        generator_version=generator_version,
+        is_monorepo=is_monorepo,
+    )
+
+
+def __filter_qualified_commit(paths: Dict[str, str], commit: Commit) -> (Commit, str):
+    """
+    Returns a tuple of a commit and libray_name.
+    A qualified commit means at least one file changes in that commit is
+    within the versioned proto_path in paths.
+
+    :param paths: a mapping from versioned proto_path to library_name.
+    :param commit: a commit under consideration.
+    :return: a tuple of a commit and library_name if the commit is
+    qualified; otherwise an empty tuple.
+    """
+    for file in commit.stats.files.keys():
+        versioned_proto_path = find_versioned_proto_path(file)
+        if versioned_proto_path in paths:
+            return commit, paths[versioned_proto_path]
+    return ()
+
+
+def __combine_commit_messages(
+    latest_commit: str,
+    baseline_commit: str,
+    commits: Dict[Commit, str],
+    generator_version: str,
+    is_monorepo: bool,
+) -> str:
+    messages = [
+        f"This pull request is generated with proto changes between googleapis commit {baseline_commit} (exclusive) and {latest_commit} (inclusive).",
+        "Qualified commits are:",
+    ]
+    for commit in commits:
+        short_sha = commit.hexsha[:7]
+        messages.append(
+            f"[googleapis/googleapis@{short_sha}](https://github.com/googleapis/googleapis/commit/{commit.hexsha})"
+        )
+
+    messages.append("BEGIN_COMMIT_OVERRIDE")
```

Comment:
What's the rationale to use BEGIN_COMMIT_OVERRIDE? It's for when we manually modify the wrong commit message for release notes.

Reply:
I think the benefits of this is to create separate entries in release note.
For example:
```
COMMIT_OVERRIDE

feat: [document-ai] expose model_type in v1 processor, so that user can see the model_type after get or list processor version

feat: [document-ai] add model_type in v1beta3 processor proto

feat: Regenerate with the Java code generator (gapic-generator-java) v2.34.0

END_COMMIT_OVERRIDE
```
will appear in release note as:
```
Features
- [document-ai] expose model_type in v1 processor, so that user can see the model_type after get or list processor version (xxx)(yyy)
- [document-ai] add model_type in v1beta3 processor proto (xxx)(yyy)
- Regenerate with the Java code generator (gapic-generator-java) v2.34.0 (xxx)(yyy)
```

Reply:
Both "nested commits" and "override" commits work. The nested one was to assist automations like owl-bot which append commits to the end of a PR. It allows for having the extended commit message while not breaking the multi-commit message parsing.

Example:

```
feat: some feature

some extended description
```

```
fix: another fix

more context about the fix
```

If you use override, you cannot keep the extended description of the first (if you add non-conventional-commit message lines between the feat and fix, then further commit messages above will be ignored:

```
BEGIN_COMMIT_OVERRIDE
feat: some feature
fix: another fix
END_COMMIT_OVERRIDE
```

If you use nested commits

```
BEGIN_NESTED_COMMIT
feat: some feature

some extended description
END_NESTED_COMMIT
BEGIN_NESTED_COMMIT
fix: another fix

more context about the fix
END_NESTED_COMMIT
```


You can even do an override with nested commits inside.

Reply:
I've changed to `BEGIN_NESTED_COMMIT` and `END_NESTED_COMMIT`.